src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkRepositoryContext.java

```diff
@@ -620,6 +631,42 @@ public void enforceLabelAttributes()
     }
   }
 
+  @StarlarkMethod(
+      name = "watch_dir",
+      doc =
+          "Tells Bazel to watch for changes to the list of entries in the given directory. Any "
+              + "changes to the entries in the directory will invalidate this repository, and "
+              + "cause it to be refetched next time.<p>\"Changes\" include entry creation, "
+              + "deletion, and renaming. Only directories outside the current Bazel workspace can "
+              + "be watched.",
+      parameters = {
+        @Param(
+            name = "path",
+            allowedTypes = {
+              @ParamType(type = String.class),
+              @ParamType(type = Label.class),
+              @ParamType(type = StarlarkPath.class)
+            },
+            doc = "path of the directory to watch."),
+      })
+  public void watchDir(Object pathO) throws EvalException, InterruptedException {
+    StarlarkPath p = getPath("watch_dir()", pathO);
+    Path path = p.getPath();
+    if (!p.isDir()) {
+      throw Starlark.errorf("watch_dir() can only be used with directories, not %s", p);
+    }
+    if (path.startsWith(directories.getWorkspace())
```

Comment:
Any reason we have to enforce this check?

Reply:
it was mostly laziness 😅 but since I moved this to `path.readdir()`, it's no longer feasible to be lazy. So I did a big refactor and now this has similar semantics as `rctx.watch()`.